### PR TITLE
Feature : Debounce on Search

### PR DESCRIPTION
**Pull request by ich-wtag**
**Overview**
[**Bonus task-2]**(https://trello.com/c/GdRu8FUF/2-bonus-task-2-debounce)
This PR includes debounce functionality while searching data. The data is persistent after refreshing.

**Types Of Change**

- [x] data filtration delayed based on search value,
- [x] while refreshing data is shown without delaying

Review
Please have a close look on this PR and kindly put your valuable comments.

[Feature__Debounce.webm](https://github.com/ich-wtag/Todo-App_Vanilla_Js/assets/153156568/da6bd7a9-fc83-424d-a3db-efac345ba330)

### CODE REVIEW

File: js/script.js [hunk 1]
```diff
@@ -184,7 +185,23 @@ const markDoneTodoHandler = (inputElement, todo) => {
     renderTodos();
 };
 
-const searchHandler = (searchedValue) => {
+const debounce =(func, timeout = 500) => {
```
Comment (hhu-wtag) [inline js/script.js]: Let's move this function to `utility.js`

File: js/script.js [hunk 2]
```diff
@@ -184,7 +185,23 @@ const markDoneTodoHandler = (inputElement, todo) => {
...
+    let timer;
+    return () => {
+        clearTimeout(timer);
+        timer = setTimeout(() => {
+            func();
+        }, timeout);
```
Comment (hhu-wtag) [inline js/script.js]: Will this work if the function had arguments?

Comment (ich-wtag) [reply js/script.js]: no bhai. when we will have arguments, we have to add arguments to the function using apply or bind method. But currently, I don't have any arguments. for that I have implemented in this way.

if we had arguments, we had to implement debounce in this way. And we have to use debounce function as declaration not the function expression.

` return (...args) => {
        clearTimeout(timer);
        timer = setTimeout(() => {
            func.apply(this, args);
        }, timeout);`

Comment (hhu-wtag) [reply js/script.js]: `debounce` is a generic function. Implementing it thinking about only current project scope is a foot gun. Cause we don't know if the function is going to take arguments in the future scope or not. If it does, then it'll lead to unexpected bug.

Comment (ich-wtag) [reply js/script.js]: okay bhai I am resolving it.

**Review by hhu-wtag — APPROVED** [review]

**Pull request merged**